discovery.md

```diff
@@ -101,18 +101,295 @@ be considered an event subscriber and a target user of this API.
+
+### Entities in the API
+
+1. `Provider`
+2. `Type`
+3. `Producer`
+
+The `Provider` and `Type` entities form the basis of the directory and can be
+used to build user experiences around discovery. The `Producer` entity is
+keyed off of `{Provider.Name, Type.Name}` and provides the necessary details to
+create a subscription for events of that `ce-type` from the selected `provider`.
+
+### REST Paths
+
+Each path in the REST API represents either represents a list (or search) over
+and entity class, or the retrieval of an individual entity by name. Each of
+these operations MUST be supported by compliant discovery providers.
+
+#### `/types?matching=[search term]`
+
+Returns a list of all types in the discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against `type`
+  names. The parameter can match any portion of the type name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"com.storage.object"`
+    * matches:
+      * `"com.storage.object.create"`
+      * `"com.storage.object.delete"`
+      * `"com.storage.object.update"`
+
+##### Returns
+
+A JSON array of `type` entities.
+
+#### `/types/{name}`
+
+Retrieves an individual type entity by exact match on the `type` name.
+
+##### Returns
+
+A `type` entity as a JSON object.
+
+#### `/types/{name}/providers`
+
+Retrieves the details about providers that offer the specified `type`.
+
+##### Returns
+
+A JSON array of `provider` entities.
+
+#### `/providers?matching=[search term]`
+
+Returns a list of all providers int he discovery system, optionally filtering on
+a provided search term (`matching`).
+
+##### matching
+
+* Type: `string`
+* Description: Search term that provides case insensitive match against
+  `provider` names. The parameter can match any portion of the provider name.
+* Constraints:
+  * OPTIONAL
+  * If present, must be non-empty
+* Examples:
+  * `"storage"`
+    * matches:
+      * `"Awesome Cloud Storage"`
+      * `"File storage system"`
+      * `"storage Storage STORAGE"`
+
+##### Returns
+
+A JSON array of `provider` entities.
+
+#### `/providers/{name}`
+
+Retrieves an individual provider entity by exact match on the `provider` name.
+
+##### Returns
+
+A `provider` entity as a JSON object.
+
+#### `/providers/{name}/types`
+
+Retrieves the details about the types that are offered by the specified
+`provider`.
+
+##### Returns
+
+A JSON array of `type` entities
+
+#### `/producer/{provider.name}/{type.name}`
+
+Retrieves the `producer` entity that specifies the information necessary to
+create subscriptions and to consume the events. The `provider.name` and
+`type.name` items in the request path correspond to the information that can
+be obtained via the `/provider` and `/type` API calls.
+
+##### Returns
+
+A `producer` entity as a JSON object.
+
+### Entity Specifications
```

Comment:
I know we're a bit early in converging on a model we all like here, but regardless, it seems like it is going to introduce concepts which aren't part of the current cloud events spec, but might be of interest to event subscribers. Might be worth thinking about exposing new cloud event attributes which enclose a reference to them (or extension attributes if discovery is optional). For example, a link to the type URL, provider URL, etc.

Some of these entities aren't part of the current cloud events spec. Thoughts on whether they might be useful to link in a cloud event instance? For example, we have a "type" identifier, but we're introducing

Reply:
For some of those things, separation does make sense. There are items that are important the discovery / subscription configuration state that aren't needed on actual events. I think provider URL is sone of those.